### ISSUE TITLE
Is it possible to use WebGPU? 

### ISSUE THREAD

**Issue by BeezBumba** [nbarkhina/N64Wasm#58]
I found an article that talks about the conversion from WebGL to WebGPU. WebGPU is said to have compute shaders and geometry shaders. Geometry shaders are what the emulator needs because, with them, every triangle has a shader that finds the right pixel colors. This explains why there is no color in the images. I posted this in a new issue since the other one was pretty long.
![Screenshot 2024-06-13 175229](https://github.com/user-attachments/assets/747a68ed-09f7-4912-9cd7-0f0ac9d7b67c)
![Screenshot 2024-06-13 174826](https://github.com/user-attachments/assets/80772d2d-99bd-4e1a-9a32-d47095a2c7ee)

**Comment by BeezBumba** [comment]
Sorry I forgot the actual link. https://developer.chrome.com/blog/from-webgl-to-webgpu

**Comment by BeezBumba** [comment]
The WebGPU using a different shading language than WebGL, but the concepts are fairly the same.

**Comment by BeezBumba** [comment]
I saw shader_frag and shader_vert files in the code directory. WebGL can't support geometry shaders. so could there have been a shader_geo.hlsl file? That might be easier if it's possible.

**Comment by nbarkhina** [comment]
those shaders are just used for the emulator overlay menu - "New Game, Save State, etc..." it's not used by the emulator itself

**Comment by BeezBumba** [comment]
So what is your option on the whole conversion? Do you think it's possible at all?

**Comment by nbarkhina** [comment]
My knowledge of shaders unfortunately is quite basic - so I would really not know

**Comment by BeezBumba** [comment]
Its not to much about shaders. It's just the conversion from WebGL to WebGPU.

**Comment by BeezBumba** [comment]
Well can I ask what spots you changed so the emulator uses webGL?

**Comment by BeezBumba** [comment]
and the spots where you are using the webgl context

**Comment by nbarkhina** [comment]
The emulator uses WebGL because that is what emscripten translates to when compiling from OpenGL. See this article - 

https://emscripten.org/docs/porting/multimedia_and_graphics/OpenGL-support.html

Here in the code is where it sets up the OpenGL Context -

https://github.com/nbarkhina/N64Wasm/blob/4aacfded69cdba1c19a07bfe7f3eb13c22252f7f/code/mymain.cpp#L632

So I don't actually use any WebGL commands, it's just straight up OpenGL which the compiler at a lower level turns into WebGL. I'm not sure how you would get it to use something else, the emulator would have to be rewritten to use something other than OpenGL - which in Desktop could use different backends (Vulkan, Direct3D, etc...) but on the web you don't have those options. So don't really think this is possible in my opinion.

**Comment by BeezBumba** [comment]
Do you know how to disable the emulator emulating lag?

**Comment by BeezBumba** [comment]
Emulators do that just for the pure accuracy.

**Comment by nbarkhina** [comment]
I don't believe there is any lag, the main tool for maintaining correct framerate is by syncing to the refresh rate of the monitor. this is done using the emscripten_set_main_loop() function:

https://github.com/nbarkhina/N64Wasm/blob/4aacfded69cdba1c19a07bfe7f3eb13c22252f7f/code/mymain.cpp#L734

you can read about it here

https://emscripten.org/docs/api_reference/emscripten.h.html

one thing you can experiment with is setting the variable doubleSpeed to true here -

https://github.com/nbarkhina/N64Wasm/blob/4aacfded69cdba1c19a07bfe7f3eb13c22252f7f/code/mymain.cpp#L42

that will attempt to run 2 frames for every 1 frame of time, however on slow machines it won't be able to keep up and won't yield any benefit.

**Comment by BeezBumba** [comment]
I asked this because some games on Ipad are a lot better now with vertex buffer objects. but that update still did something with safari.

**Comment by BeezBumba** [comment]
new update: the parallel graphics plugin is actually the inaccurate one. I tested it on retroarch and it gave me the same problems.

**Comment by BeezBumba** [comment]
we could try using gliden plus instead

**Comment by BeezBumba** [comment]
I found this one on gitlab https://gitlab.com/parallel-launcher/parallel-n64

**Comment by nbarkhina** [comment]
interesting, ok good to know it wasn't something specific to the web version.

**Comment by BeezBumba** [comment]
It says it’s used specifically for the parallel launcher and that one works fine. So should you try it?

**Comment by nbarkhina** [comment]
I used the one that was already built into the parallel core in retroarch. Not sure how this one is different or if it's feasible to integrate it but if I have time I'll take a look

**Comment by BeezBumba** [comment]
Should I add anything for the new mupen64plus-video-gliden64 plugin code wise? Wondering what this video plugin might offer.

**Comment by nbarkhina** [comment]
You can try experimenting with if you like, I unfortunately have not had time to look at it been busy with work

**Comment by nbarkhina** [comment]
I've been working on getting the Rice plugin working - stay tuned I will link to a version you can try and let me know if it fixes the issue.

**Comment by BeezBumba** [comment]
Nice! can't wait to see that!

**Comment by nbarkhina** [comment]
ok give this page a try -

https://neilb.net/n64beta/

**Comment by BeezBumba** [comment]
It works!
![Screenshot 2024-11-10 195133](https://github.com/user-attachments/assets/5aa4ed78-654b-43af-af19-7ec7140d966b)

**Comment by BeezBumba** [comment]
the ongoing visual bugs are very minor and not as big so thats good.

**Comment by BeezBumba** [comment]
can you link the code?

**Comment by nbarkhina** [comment]
I haven't published the code yet I'm still cleaning it up a bit. I should have it published in a few days and will link you to it as soon as it's available.

**Comment by nbarkhina** [comment]
ok here is the checkin with the code - 

https://github.com/nbarkhina/N64Wasm/commit/3056c89d73efea037aa5653a065ab65947e00149

**Comment by BeezBumba** [comment]
There is this thing where the whole game covers about 1/4 of the screen it might be a resolution issue?
![Screenshot 2024-11-12 184744](https://github.com/user-attachments/assets/cb6c3272-5368-4801-80c1-1546dc9e596d)

**Comment by BeezBumba** [comment]
It might be lines 390 and 391 in RiceConfig.cpp i have to change to global width/height?
```
screen_width = 640;
   screen_height = 480;
```

**Comment by nbarkhina** [comment]
Yes try updating those lines
